### PR TITLE
HS-1285: Add filter to list alerts

### PR DESCRIPTION
**Pull request by BenjaminJ**
o add severity filter
o add time filter ('today', '24h', '7d')
o add severity and time filter
o update cucumber test

HS-1285

## Description
Add filter to list alerts
module:
alert
rest-server

## Jira link(s)
- https://issues.opennms.org/browse/HS-1285

## Checklist
* [ ] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts

### CODE REVIEW

File: alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEngine.java [hunk 1]
```diff
@@ -116,12 +116,12 @@ public synchronized void handleDeletedAlert(Alert alert) {
...
     private synchronized void tick() {
-        LOG.info("Tick with: {}", alertsByReductionKeyByTenantId);
-        // Delete alerts more than 1 hour old
+        LOG.debug("Tick with: {}", alertsByReductionKeyByTenantId);
+        // Delete alerts more than 2 weeks old
         alertsByReductionKeyByTenantId.forEach((tenantId, alertsByReductionKey) -> {
             Context.current().withValue(GrpcConstants.TENANT_ID_CONTEXT_KEY, tenantId).run(()-> {
                 alertsByReductionKey.values().stream()
-                    .filter(a -> a.getLastUpdateTimeMs() < (System.currentTimeMillis() - TimeUnit.HOURS.toMillis(1)))
+                    .filter(a -> a.getLastUpdateTimeMs() < (System.currentTimeMillis() - TimeUnit.DAYS.toMillis(14)))
```
Comment (yangli2014) [inline alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEngine.java:124]: We should avoid hardcoded values.

File: alert/src/main/java/org/opennms/horizon/alertservice/db/repository/AlertRepository.java [hunk 2]
```diff
@@ -28,11 +28,23 @@
...
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface AlertRepository extends JpaRepository<Alert, Long> {
     Alert findByReductionKey(String reductionKey);
+
+    Page<Alert> findBySeverityIn(List<Severity> severity, Pageable pageable);
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/db/repository/AlertRepository.java:45]: I hope you are aware of the fact that tenantId is assigned implicitly here.  If these repository calls are made outside grpc api, we need to have context around it.

It was decided that we will get rid of this implicit handing of tenantId

File: alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java [hunk 3]
```diff
@@ -50,22 +58,42 @@
...
-        String nextPageToken = !request.getNextPageToken().isEmpty() ? request.getNextPageToken() : PAGE;
-
-        // Create a PageRequest object based on the page size and next page token
-        PageRequest pageRequest = PageRequest.of(Integer.parseInt(nextPageToken), pageSize);
-
-        Page<org.opennms.horizon.alertservice.db.entity.Alert> alertPage = alertRepository.findAll(pageRequest);
+        int pageSize = request.getPageSize() != 0 ? request.getPageSize() : PAGE_SIZE_DEFAULT;
+        String nextPageToken = !request.getNextPageToken().isEmpty() ? request.getNextPageToken() : PAGE_DEFAULT;
+
+        String filter = !request.getFilter().isEmpty() ? request.getFilter() : "";
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java:74]: Is this different than `String filter = request.getFilter()`

File: alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java [hunk 4]
```diff
@@ -50,22 +58,42 @@
...
+        String sortBy = !request.getSortBy().isEmpty() ? request.getSortBy() : SORT_BY_DEFAULT;
+        boolean sortAscending = request.getSortAscending();
+
+        // Create a PageRequest object based on the page size, next page token, filter, and sort parameters
+        Sort.Direction sortDirection = sortAscending ? Sort.Direction.ASC : Sort.Direction.DESC;
+        PageRequest pageRequest = PageRequest.of(Integer.parseInt(nextPageToken), pageSize, Sort.by(sortDirection, sortBy));
+        Date[] timeRange;
+        Page<org.opennms.horizon.alertservice.db.entity.Alert> alertPage =
+            switch (filter) {
+                case "severity":
```
Comment (cgorantla) [inline alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java:85]: should we make a Enum on all of the supported filters

File: alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEngine.java (hunk 1 above)
Comment (BenjaminJ) [reply alert/src/main/java/org/opennms/horizon/alertservice/service/AlertEngine.java:124]: Fix in https://github.com/OpenNMS-Cloud/horizon-stream/pull/913

File: alert/src/main/java/org/opennms/horizon/alertservice/db/repository/AlertRepository.java (hunk 2 above)
Comment (BenjaminJ) [reply alert/src/main/java/org/opennms/horizon/alertservice/db/repository/AlertRepository.java:45]: Addressed in https://github.com/OpenNMS-Cloud/horizon-stream/pull/913

File: alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java (hunk 3 above)
Comment (BenjaminJ) [reply alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java:74]: Refactored in https://github.com/OpenNMS-Cloud/horizon-stream/pull/913

File: alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java (hunk 4 above)
Comment (BenjaminJ) [reply alert/src/main/java/org/opennms/horizon/alertservice/grpc/AlertGrpcService.java:85]: Refactored in https://github.com/OpenNMS-Cloud/horizon-stream/pull/913

**Pull request merged**